### PR TITLE
refactor cost read-in and cost manipulation - moving all cost related codes to one place, improve documentation

### PR DESCRIPTION
**Pull request by cchrisgong**
## Purpose of this PR
refactor cost read in and manipulation part of the core/datainput.gms, so all of the code are close together in one place

## Type of change

(Make sure to delete from the Type-of-change list the items not relevant to your PR)

- [ ] Bug fix 
- [ X] Refactoring
- [ ] New feature 
- [ ] Minor change (default scenarios show only small differences)
- [ ] Fundamental change
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the [coding etiquette](https://github.com/remindmodel/remind/blob/develop/main.gms#L80)
- [x] I performed a self-review of my own code
- [x] I explained my changes within the PR, particularly in hard-to-understand areas
- [x] I checked that the [in-code documentation](https://github.com/remindmodel/remind/blob/develop/main.gms#L120) is up-to-date
- [x] I adjusted the reporting in [`remind2`](https://github.com/pik-piam/remind2) where it was needed
- [x] I adjusted `forbiddenColumnNames` in [readCheckScenarioConfig.R](https://github.com/remindmodel/remind/blob/develop/scripts/start/readCheckScenarioConfig.R) in case the PR leads to deprecated switches
- [x] All automated model tests pass (`FAIL 0` in the output of `make test`)
- [x] The changelog `CHANGELOG.md` [has been updated correctly](https://gitlab.pik-potsdam.de/rse/rsewiki/-/wikis/Standards-for-Writing-a-Changelog)

## Further information (optional):

* Test runs are here: 
* before (develop /trunk): /p/tmp/chengong/remind_costrefac/remind/output/testOneRegi_develop_2024-02-19_15.34.40
* after refactoring: /p/tmp/chengong/remind_costrefac/remind/output/testOneRegi_cost_2024-02-19_15.30.19
* Comparison of results (what changes by this PR?):

### CODE REVIEW

File: core/datainput.gms [hunk 1]
```diff
@@ -199,11 +230,190 @@ fm_dataglob("incolearn",te)          = s_D2015_2_D2005 * fm_dataglob("incolearn"
...
+p_inco0(ttot,regi,te)               = s_DpKW_2_TDpTW       * p_inco0(ttot,regi,te);
+
 *RP* rescale the global CSP investment costs in REMIND: Originally we assume a SM3/12h setup, while the cost data from IEA for the short term seems rather based on a SM2/6h setup (with 40% average CF)
 *** Accordingly, also decrease long-term costs in REMIND to 0.7 of the current values
 fm_dataglob("inco0","csp")              = 0.7 * fm_dataglob("inco0","csp");
 fm_dataglob("incolearn","csp")          = 0.7 * fm_dataglob("incolearn","csp");
 
+
+*** --------------------------------------------------------------------------------
+****** Regionalize investment cost data
```
Comment (robertpietzcker) [inline core/datainput.gms:248]: please make sure there are never more than 3 asterisks behind each other

File: core/datainput.gms [hunk 2]
```diff
@@ -130,14 +130,62 @@ pm_shGasLiq_fe_up(ttot,regi,sector)=0;
...
+parameter p_inco0(ttot,all_regi,all_te)     "regionalized technology costs Unit: USD$/KW"
+/
+$ondelim
+$include "./core/input/p_inco0.cs4r"
+$offdelim
+/
+;
+
+*** initializing energy service capital
+pm_esCapCost(tall,all_regi,all_teEs) = 0;
```
Comment (robertpietzcker) [inline core/datainput.gms:182]: does energy service capital still exist in the latest REMIND version, after transport|complex and buildings|putty were deleted?

**Review by robertpietzcker — APPROVED** [review]
I think it looks ok, but with this amount of code shifted, there could be errors. 

I would propose you merge this to prevent merge conflicts, but do additional tests, namely compare all the parameters that might be involved (pm_data, p_inco0, pm_inco0_t, xxx)

I guess the easiest way would be to write them into a file in both the old and the new version, and do a quick winmerge/diff to see if they are exactly the same. 
Please do this for both global and regionalized tech costs, and SSP1, SSP2 and SSP5

Then add a link to the folder where the old/new parameter results are saved

It would be painful if one of them changes without us being aware :-)

File: core/datainput.gms (hunk 2 above)
Comment (cchrisgong) [reply core/datainput.gms:182]: there are sill some 
image
@robertpietzcker

**Comment by robertpietzcker** [comment]
before we forget about this - did you do the comparison of the parameters before/after for the different model configurations?

**Comment by cchrisgong** [comment]
I did a test one regi in test/scenconfig_quick.csv, I believe it is ssp2? And compared endogenous parameters of capital cost. I'm submitting a patch though on some small things.

Sorry for the big change, but I think it shouldn't affect models

**Comment by robertpietzcker** [comment]
No, I was talking about my comment in my review: 
"but do additional tests, namely compare all the parameters that might be involved (pm_data, p_inco0, pm_inco0_t, xxx)

I guess the easiest way would be to write them into a file in both the old and the new version, and do a quick winmerge/diff to see if they are exactly the same.
Please do this for both global and regionalized tech costs, and SSP1, SSP2 and SSP5

Then add a link to the folder where the old/new parameter results are saved"

 so not a "are the results the same in a testOneRegi", but comparing the full parameters after datainput. if you do it in a file (simply write into a file with "put"), it will take 1 click on winmerge to check if they are identical; do this for all the setups (global tech costs; regional tech costs; SSP1; SSP2, SSP5) and then we know everything is 100% correct and we never need to worry if something indeed might have changed. 

this is much clearer than doing a REMIND run, because the model is not deterministic and so it will be more challenging to decide "are these minimal differences just a slightly different optimum, or is something actually different in the model"

**Comment by cchrisgong** [comment]
I see, thanks, yes, I think I'll do that till Friday if that's okay. I'll submit a short patch PR just for the obvious things. I haven't used put before, I'm also in Linux system. So it will take me a day or two to get there, I hope that's okay

**Comment by robertpietzcker** [comment]
sure!
on linux, diff will work as well (or better?) than winmerge :-)

and the put utility is quite straight forward, if I remember correctly.

Thanks!

**Pull request merged**